lang/LangPrimSource/PyrArrayPrimitives.cpp

```diff
@@ -2474,6 +2477,95 @@ int prArrayUnlace(struct VMGlobals* g, int numArgsPushed) {
     return errNone;
 }
 
+int arrayLevenshteinDistance(struct VMGlobals* g, PyrSlot* result, PyrObject* thisArray, PyrObject* thatArray) {
+    int distance = sc_max(thisArray->size, thatArray->size); // assume they're completely different
```

Comment:
this was actually to avoid early exits, following 'return once' philosophy.

if multiple returns are preferred, I can change it - but I personally prefer the readability of return once.
what do you think?

Reply:
you already have multiple returns in this function, though -- 5 of them :)

i find more and more that i like multiple returns when some conditions are met that specifically apply here:
1. the function can be written so that there's no return at the end, in which case you know every branch results in a return (the compiler would complain if it didn't)
2. a switch block can be written so every case results in a return -- you can eliminate the `break` statements and save some space that way, and it's easy to visually scan down the cases and see each return because they're all at the same indentation level and only 2 or 3 lines apart.

but this is just personal style, and it changes over time.

the reason i initially made this comment was because i was reading through and couldn't figure out quickly if this original value of `distance` was being used anywhere, and then thought it might be a programming error or out of date code when i realized it wasn't.

the tradeoff between using or not using early exits here is between more complex control flow vs more complex state management. if you have early exits like i was suggesting, you make the function a bit less stateful because i don't have to track the value of `distance` as i read through it. but the control flow is a little more complicated as a result. what i was saying above is that i've personally internalized some control flow schemas like "early exits from every case of a `switch`", and so i find them easy to understand at a glance.

Reply:
Thanks Brian, I've been trying to work through this one but wanted to check the implementation first.

Currently, `distance` is used to set the value on the stack, and the return value of `arrayLevenshteinDistance` is the error code for the primitive.

If `arrayLevenshteinDistance` returns the distance, from the switch, then capturing the error code becomes more fiddly. The distance should really be a `size_t` (currently an `int`), so negative distances can't indicate failure. This also raises the question of how to cast the distance into something sclang can handle?

At this point, I can either return a `std::pair<size_t, int>` of distance and error code, or pass in an `int&` as a parameter to `arrayLevenshteinDistance` for it to set an error code if needed.

For the first, every return statement in the switch needs an added initialiser list, not very clean.
For the second, the error code could be set to `errNone` at the beginning, and only in cases of error adding an extra line to set it to `errWrongType` - but the question becomes, what does `arrayLevenshteinDistance` return in this case? 0? Max array size?

I'm leaning towards the second option, but wanted to check first. Minimal impl below to help explain..

Thanks!

```cpp
// current
int arrayLevenshteinDistance(PyrSlot*, PyrObject*, PyrObject*);

int prArrayLevenshteinDistance() {
    int errCode = arrayLevenshteinDistance(resultPtr, thisArray, thatArray);
    return errCode;
}

// A --
std::pair<size_t, int> arrayLevensteinDistance(PyrObject*, PyrObject*);

int prArrayLevenshteinDistance() {
    auto distResult = arrayLevenshteinDistance(thisArray, thatArray);

    if (distResult.second == errNone)
        SetInt(resultPtr, distResult.first);

    return distResult.second;
}

// B --
size_t arrayLevenshteinDistance(PyrObject*, PyrObject*, int& errorCode);

int prArrayLevenshteinDistance() {
    int errCode = errNone;
    size_t distance = arrayLevenshteinDistance(thisArray, thatArray, &errCode);

    if (errCode == errNone)
        SetInt(resultPtr, distance);

    return errCode;
}
```